momaland/envs/multiwalker_stability/momultiwalker_stability_base.py

```diff
@@ -0,0 +1,228 @@
+        self.forward_reward = forward_reward
+        self.fall_reward = fall_reward
+        self.terminate_reward = terminate_reward
+        self.stability_reward = stability_reward
+        self.terrain_length = terrain_length
+        self.terrain_step = terrain_step
+
+    @property
+    def reward_space(self):
+        """Reward space shape = 2 element 1D array, each element representing 1 objective.
+
+        1. scalarized reward of: package moving forward, no walkers falling, and package not falling
+        2. package not tipping
+        """
+        return spaces.Box(
+            low=np.array([-210, -1.75]),
+            high=np.array([0.46, 0]),
+            shape=(2,),
+            dtype=np.float32,
+        )
+
+
+class MOMultiWalkerStabilityEnv(pz_multiwalker_base):
+    """Multiwalker problem domain environment engine.
+
+    Deals with the simulation of the environment.
+    """
+
+    @override
+    def __init__(
+        self,
+        n_walkers=3,
+        position_noise=1e-3,
+        angle_noise=1e-3,
+        forward_reward=1.0,
+        terminate_reward=-100.0,
+        fall_reward=-10.0,
+        stability_reward=-100,
+        shared_reward=True,
+        terminate_on_fall=True,
+        remove_on_fall=True,
+        terrain_length=TERRAIN_LENGTH,
+        max_cycles=500,
+        render_mode=None,
+    ):
+        self.stability_reward = stability_reward
+        self.previous_pkg_angle = 0
+        super().__init__(
+            n_walkers=n_walkers,
+            position_noise=position_noise,
+            angle_noise=angle_noise,
+            forward_reward=forward_reward,
+            terminate_reward=terminate_reward,
+            fall_reward=fall_reward,
+            shared_reward=shared_reward,
+            terminate_on_fall=terminate_on_fall,
+            remove_on_fall=remove_on_fall,
+            terrain_length=terrain_length,
+            max_cycles=max_cycles,
+            render_mode=render_mode,
+        )
+        self.setup()
+        self.last_rewards = [np.zeros(shape=(2,), dtype=np.float32) for _ in range(self.n_walkers)]
+
+    @override
+    def setup(self):
+        """Continuation of the `__init__`."""
+        super().setup()
+        init_y = TERRAIN_HEIGHT + 2 * LEG_H
+        self.walkers = [
+            MOBipedalWalkerStability(
+                self.world,
+                self.forward_reward,
+                self.fall_reward,
+                self.terminate_reward,
+                self.stability_reward,
+                init_x=sx,
+                init_y=init_y,
+                seed=self.seed_val,
+            )
+            for sx in self.start_x
+        ]
+        self.reward_space = [agent.reward_space for agent in self.walkers]
+
+    @override
+    def reset(self):
+        obs = super().reset()
+        self.last_rewards = [np.zeros(shape=(2,), dtype=np.float32) for _ in range(self.n_walkers)]
+        self.previous_pkg_angle = 0
```

Comment:
same here

Reply:
self.package is created in reset(). I need to keep it here.